src/elevenlabs/conversation.py

```diff
@@ -0,0 +1,282 @@
+        client: BaseElevenLabs,
+        agent_id: str,
+        *,
+        requires_auth: bool,
+        audio_interface: AudioInterface,
+        callback_agent_response: typing.Optional[typing.Callable[[str], None]] = None,
+        callback_user_transcript: typing.Optional[typing.Callable[[str], None]] = None,
+        callback_latency_measurement: typing.Optional[typing.Callable[[int], None]] = None,
+    ):
+        """Conversational AI session.
+
+        BETA: This API is subject to change without regard to backwards compatibility.
+
+        Args:
+            client: The ElevenLabs client to use for the conversation.
+            agent_id: The ID of the agent to converse with.
+            requires_auth: Whether the agent requires authentication.
+            audio_interface: The audio interface to use for input and output.
+            callback_agent_response: Callback for agent responses.
+            callback_user_transcript: Callback for user transcripts.
+            callback_latency_measurement: Callback for latency measurements (in milliseconds).
+        """
+
+        self.client = client
+        self.agent_id = agent_id
+        self.requires_auth = requires_auth
+
+        self.audio_interface = audio_interface
+        self.callback_agent_response = callback_agent_response
+        self.callback_user_transcript = callback_user_transcript
+        self.callback_latency_measurement = callback_latency_measurement
+
+    def start_session(self):
+        """Starts the conversation session.
+
+        Will run in background thread until `end_session` is called.
+        """
+        ws_url = self._get_signed_url() if self.requires_auth else self._get_wss_url()
+        self._thread = threading.Thread(target=self._run, args=(ws_url,))
+        self._thread.start()
+
+    def end_session(self):
+        """Ends the conversation session."""
+        self.audio_interface.stop()
+        self._should_stop.set()
+
+    def wait_for_session_end(self) -> typing.Optional[str]:
+        """Waits for the conversation session to end.
+
+        You must call `end_session` before calling this method, otherwise it will block.
+
+        Returns the conversation ID, if available.
+        """
+        if not self._thread:
+            raise RuntimeError("Session not started.")
+        self._thread.join()
+        return self._conversation_id
+
+    def _run(self, ws_url: str):
+        with connect(ws_url) as ws:
+
+            def input_callback(audio):
+                ws.send(
+                    json.dumps(
+                        {
+                            "user_audio_chunk": base64.b64encode(audio).decode(),
+                        }
+                    )
+                )
+
+            self.audio_interface.start(input_callback)
+            while not self._should_stop.is_set():
+                try:
+                    message = json.loads(ws.recv(timeout=0.5))
+                    self._handle_message(message, ws)
```

Comment:
check for is_set here will prohibit message parsing past `stop()`

Reply:
Well that is the idea, `end_session` will close the websocket and the audio streams. Do you want some grace period here or what is the concern?

Reply:
I would add the check right before handle message again. 

Time 0: while not is_set() -> True
Time 0.1: stop()
Time 0.5: ws.recv() returns message
Time 0.501:  message is still handled

Reply:
Ah I see you want the opposite. OK added even though it probably doesn't matter that much.